### PR TITLE
ARROW-4250: [C++] adding explicit epsilon for ApproxEquals and corresponding assert macro

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

**Comment by bkietz** [comment]
@wesm are there any other assertions which should be replaced with the approximate check?

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -275,11 +275,15 @@ class ARROW_EXPORT Array {
...
+  /// Equality comparison with another array
   bool Equals(const Array& arr) const;
   bool Equals(const std::shared_ptr<Array>& arr) const;
 
-  bool ApproxEquals(const std::shared_ptr<Array>& arr) const;
-  bool ApproxEquals(const Array& arr) const;
+  /// Approximate equality comparison with another array
+  ///
+  /// epsilon is only used if this is FloatArray or DoubleArray
+  bool ApproxEquals(const std::shared_ptr<Array>& arr, double epsilon = 1E-5) const;
```
Comment (fsaintjacques) [inline cpp/src/arrow/array.h]: Make the default a static constexpr `kDefaultEpsilon`.

File: cpp/src/arrow/compare.cc [hunk 2]
```diff
@@ -523,24 +524,23 @@ class ArrayEqualsVisitor : public RangeEqualsVisitor {
...
   using T = typename TYPE::c_type;
 
   const T* left_data = left.raw_values();
   const T* right_data = right.raw_values();
 
-  static constexpr T EPSILON = static_cast<T>(1E-5);
-
   if (left.null_count() > 0) {
     for (int64_t i = 0; i < left.length(); ++i) {
       if (left.IsNull(i)) continue;
```
Comment (fsaintjacques) [inline cpp/src/arrow/compare.cc]: I know this is not related to this commit, but should it check right.IsNull?

File: cpp/src/arrow/compare.h [hunk 3]
```diff
@@ -42,7 +42,8 @@ bool ARROW_EXPORT SparseTensorEquals(const SparseTensor& left, const SparseTenso
 
 /// Returns true if the arrays are approximately equal. For non-floating point
 /// types, this is equivalent to ArrayEquals(left, right)
-bool ARROW_EXPORT ArrayApproxEquals(const Array& left, const Array& right);
+bool ARROW_EXPORT ArrayApproxEquals(const Array& left, const Array& right,
+                                    double epsilon = 1E-5);
```
Comment (fsaintjacques) [inline cpp/src/arrow/compare.h]: Use the previously defined constant, or define it here, not sure.

**Review by fsaintjacques — COMMENTED** [review]
Minor nit, just a constexpr default.

**Comment by fsaintjacques** [comment]
@bkietz I'd say wait for https://github.com/apache/arrow/pull/4005 and then rebase.

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/array.h]: will do

File: cpp/src/arrow/compare.cc (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/compare.cc]: No, null bitmaps are compared in [BaseDataEquals](https://github.com/apache/arrow/blob/df562103ab9a17540b0dd8fe01842e94a78fff31/cpp/src/arrow/compare.cc#L583-L586)

**Comment by pitrou** [comment]
#4005 isn't ready yet. @bkietz, can you rebase?

**Review by pitrou — APPROVED** [review]
+1

**Pull request closed without merging**